File: app/src/main/java/com/majeur/psclient/model/Stats.java
package com.majeur.psclient.model;

import androidx.annotation.Nullable;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.Serializable;

public class Stats implements Serializable {

    public int hp;
    public int atk;
    public int def;
    public int spa;
    public int spd;
    public int spe;

    public Stats(JSONObject jsonObject) throws JSONException {
        hp = jsonObject.optInt("hp", 0);
        atk = jsonObject.getInt("atk");
        def = jsonObject.getInt("def");
        spa = jsonObject.getInt("spa");
        spd = jsonObject.getInt("spd");
        spe = jsonObject.getInt("spe");
    }

    public Stats(int hp, int atk, int def, int spa, int spd, int spe) {
        this.hp = hp;
        this.atk = atk;
        this.def = def;
        this.spa = spa;
        this.spd = spd;
        this.spe = spe;
    }

    public Stats(int defaultValue) {
        this.hp = defaultValue;
        this.atk = defaultValue;
        this.def = defaultValue;
        this.spa = defaultValue;
        this.spd = defaultValue;
        this.spe = defaultValue;
    }

    public void set(int index, int value) {
        switch (index) {
            case 0:
                hp = value;
                break;
            case 1:
                atk = value;
                break;
            case 2:
                def = value;
                break;
            case 3:
                spa = value;
                break;
            case 4:
                spd = value;
                break;
            case 5:
                spe = value;
                break;
        }
    }

    public int get(int index) {
        return toArray()[index];
    }

    public int[] toArray() {
        return new int[]{hp, atk, def, spa, spd, spe};
    }

    public static String getName(int index) {
        switch (index) {
            case 0: return "HP";
            case 1: return "Atk";
            case 2: return "Def";
            case 3: return "SpA";
            case 4: return "Spd";
            case 5: return "Spe";
            default: return null;
        }
    }

    @Override
    public String toString() {
        return "Stats{" +
                "hp=" + hp +
                ", atk=" + atk +
                ", def=" + def +
                ", spa=" + spa +
                ", spd=" + spd +
                ", spe=" + spe +
                '}';
    }

    public int sum() {
        return hp + atk + def + spa + spd + spe;
    }

    public String hpType() {
        int a = hp % 2 == 0 ? 0 : 1;
        int b = atk % 2 == 0 ? 0 : 2;
        int c = def % 2 == 0 ? 0 : 4;
        int d = spe % 2 == 0 ? 0 : 8;
        int e = spa % 2 == 0 ? 0 : 16;
        int f = spd % 2 == 0 ? 0 : 32;
        int t = (a + b + c + d + e + f) * 15 / 63;
        switch (t) {
            case 0: return "Fighting";
            case 1: return "Flying";
            case 2: return "Poison";
            case 3: return "Ground";
            case 4: return "Rock";
            case 5: return "Bug";
            case 6: return "Ghost";
            case 7: return "Steel";
            case 8: return "Fire";
            case 9: return "Water";
            case 10: return "Grass";
            case 11: return "Electric";
            case 12: return "Psychic";
            case 13: return "Ice";
            case 14: return "Dragon";
            case 15: return "Dark";
            default: return null;
        }
    }

    public void setForHpType(@Nullable String type) {
        if (type == null) type = "Dark"; // Default all 31 iv spread
        switch (type) {
            case "Bug":			hp = 31; atk = 31; def = 31; spa = 31; spd = 30; spe = 30; break;
            case "Dark":		hp = 31; atk = 31; def = 31; spa = 31; spd = 31; spe = 31; break;
            case "Dragon":		hp = 30; atk = 31; def = 31; spa = 31; spd = 31; spe = 31; break;
            case "Electric":	hp = 31; atk = 31; def = 31; spa = 30; spd = 31; spe = 31; break;
            case "Fighting":	hp = 31; atk = 31; def = 30; spa = 30; spd = 30; spe = 30; break;
            case "Fire":		hp = 31; atk = 30; def = 31; spa = 30; spd = 31; spe = 30; break;
            case "Flying":		hp = 31; atk = 31; def = 31; spa = 30; spd = 30; spe = 30; break;
            case "Ghost":		hp = 31; atk = 30; def = 31; spa = 31; spd = 30; spe = 31; break;
            case "Grass":		hp = 30; atk = 31; def = 31; spa = 30; spd = 31; spe = 31; break;
            case "Ground":		hp = 31; atk = 31; def = 31; spa = 30; spd = 30; spe = 31; break;
            case "Ice":			hp = 31; atk = 31; def = 31; spa = 31; spd = 31; spe = 30; break;
            case "Poison":		hp = 31; atk = 31; def = 30; spa = 30; spd = 30; spe = 31; break;
            case "Psychic":		hp = 30; atk = 31; def = 31; spa = 31; spd = 31; spe = 30; break;
            case "Rock":		hp = 31; atk = 31; def = 30; spa = 31; spd = 30; spe = 30; break;
            case "Steel":		hp = 31; atk = 31; def = 31; spa = 31; spd = 30; spe = 31; break;
            case "Water":		hp = 31; atk = 31; def = 31; spa = 30; spd = 31; spe = 30; break;
        }
    }

    public static boolean checkHpType(String type) {
        if (type == null) return false;
        type = type.toLowerCase().trim();
        switch (type) {
            case "bug":			return true;
            case "dark":		return true;
            case "dragon":		return true;
            case "electric":	return true;
            case "fighting":	return true;
            case "fire":		return true;
            case "flying":		return true;
            case "ghost":		return true;
            case "grass":		return true;
            case "ground":		return true;
            case "ice":			return true;
            case "poison":		return true;
            case "psychic":		return true;
            case "rock":		return true;
            case "steel":		return true;
            case "water":		return true;
            default:            return false;
        }
    }

    public static int[] calculateSpeedRange(int level, int baseSpe, String tier, int gen) {
        boolean isRandomBattle = tier.contains("Random Battle") || (tier.contains("Random") && tier.contains("Battle") && gen >= 6);

        float minNature = (isRandomBattle || gen < 3) ? 1f : 0.9f;
        float maxNature = (isRandomBattle || gen < 3) ? 1f : 1.1f;
        int maxIv = (gen < 3) ? 30 : 31;

        int min;
        int max;
        if (tier.contains("Let's Go")) {
            min = tr(tr(tr(2 * baseSpe * level / 100f + 5) * minNature) * tr((70f / 255f / 10f + 1) * 100) / 100f);
            max = tr(tr(tr((2 * baseSpe + maxIv) * level / 100f + 5) * maxNature) * tr((70f / 255f / 10f + 1) * 100) / 100f);
            if (tier.contains("No Restrictions")) max += 200;
            else if (tier.contains("Random")) max += 20;
        } else {
            float maxIvEvOffset = maxIv + ((isRandomBattle && gen >= 3) ? 21 : 63);
            min = tr(tr(2 * baseSpe * level / 100f + 5) * minNature);
            max = tr(tr((2 * baseSpe + maxIvEvOffset) * level / 100f + 5) * maxNature);
        }
        return new int[]{min, max};
    }

    private static int tr(float value) {
        return (int) value;
    }


    public static int calculateStat(int base, int iv, int ev, int niv, float nat) {
        return (int) ((((2 * base + iv + ev / 4) * niv) / 100 + 5) * nat);
    }

    public static int calculateHp(int base, int iv, int ev, int niv) {
        return ((2 * base + iv + ev / 4) * niv) / 100 + niv + 10;
    }

    public static int toIndex(String name) {
        name = name.trim().toLowerCase();
        if (name.contains("atk") || name.contains("attack")) {
            if (name.contains("sp")) return 3;
            else return 1;
        } else if (name.contains("def")) {
            if (name.contains("sp")) return 4;
            else return 2;
        } else if (name.contains("sp")) {
            return 5;
        } else if (name.contains("h")) {
            return 0;
        }
        return -1;
    }
}

File: psclient/src/main/java/com/majeur/psclient/model/common/Stats.kt
package com.majeur.psclient.model.common

import android.text.SpannableStringBuilder
import com.majeur.psclient.util.concat
import com.majeur.psclient.util.small
import com.majeur.psclient.util.toId
import org.json.JSONObject
import java.io.Serializable
import java.util.*


class Stats() : Serializable {

    var hp = 0
    var atk = 0
    var def = 0
    var spa = 0
    var spd = 0
    var spe = 0

    val array
        get() = intArrayOf(hp, atk, def, spa, spd, spe)

    constructor(json: JSONObject) : this() {
        hp = json.optInt("hp", 0)
        atk = json.getInt("atk")
        def = json.getInt("def")
        spa = json.getInt("spa")
        spd = json.getInt("spd")
        spe = json.getInt("spe")
    }

    constructor(hp: Int, atk: Int, def: Int, spa: Int, spd: Int, spe: Int) : this() {
        this.hp = hp
        this.atk = atk
        this.def = def
        this.spa = spa
        this.spd = spd
        this.spe = spe
    }

    constructor(defaultValue: Int) : this() {
        hp = defaultValue
        atk = defaultValue
        def = defaultValue
        spa = defaultValue
        spd = defaultValue
        spe = defaultValue
    }

    fun setAll(value: Int) {
        hp = value
        atk = value
        def = value
        spa = value
        spd = value
        spe = value
    }

    fun set(stats: Stats) {
        hp = stats.hp
        atk = stats.atk
        def = stats.def
        spa = stats.spa
        spd = stats.spd
        spe = stats.spe
    }

    fun set(name: String, value: Int) = set(toIndex(name), value)

    fun set(index: Int, value: Int) = when (index) {
        0 -> hp = value
        1 -> atk = value
        2 -> def = value
        3 -> spa = value
        4 -> spd = value
        5 -> spe = value
        else -> Unit
    }

    fun get(index: Int) = array[index]

    fun sum() = hp + atk + def + spa + spd + spe

    fun hpType(): String {
        val a = if (hp % 2 == 0) 0 else 1
        val b = if (atk % 2 == 0) 0 else 2
        val c = if (def % 2 == 0) 0 else 4
        val d = if (spe % 2 == 0) 0 else 8
        val e = if (spa % 2 == 0) 0 else 16
        val f = if (spd % 2 == 0) 0 else 32
        val t = (a + b + c + d + e + f) * 15 / 63
        return when (t) {
            0 -> "Fighting"
            1 -> "Flying"
            2 -> "Poison"
            3 -> "Ground"
            4 -> "Rock"
            5 -> "Bug"
            6 -> "Ghost"
            7 -> "Steel"
            8 -> "Fire"
            9 -> "Water"
            10 -> "Grass"
            11 -> "Electric"
            12 -> "Psychic"
            13 -> "Ice"
            14 -> "Dragon"
            15 -> "Dark"
            else -> ""
        }
    }

    fun setForHpType(type: String?) {
        when (type ?: "Dark") { // Dark is default all 31 iv spread
            "Bug" -> {
                hp = 31
                atk = 31
                def = 31
                spa = 31
                spd = 30
                spe = 30
            }
            "Dark" -> {
                hp = 31
                atk = 31
                def = 31
                spa = 31
                spd = 31
                spe = 31
            }
            "Dragon" -> {
                hp = 30
                atk = 31
                def = 31
                spa = 31
                spd = 31
                spe = 31
            }
            "Electric" -> {
                hp = 31
                atk = 31
                def = 31
                spa = 30
                spd = 31
                spe = 31
            }
            "Fighting" -> {
                hp = 31
                atk = 31
                def = 30
                spa = 30
                spd = 30
                spe = 30
            }
            "Fire" -> {
                hp = 31
                atk = 30
                def = 31
                spa = 30
                spd = 31
                spe = 30
            }
            "Flying" -> {
                hp = 31
                atk = 31
                def = 31
                spa = 30
                spd = 30
                spe = 30
            }
            "Ghost" -> {
                hp = 31
                atk = 30
                def = 31
                spa = 31
                spd = 30
                spe = 31
            }
            "Grass" -> {
                hp = 30
                atk = 31
                def = 31
                spa = 30
                spd = 31
                spe = 31
            }
            "Ground" -> {
                hp = 31
                atk = 31
                def = 31
                spa = 30
                spd = 30
                spe = 31
            }
            "Ice" -> {
                hp = 31
                atk = 31
                def = 31
                spa = 31
                spd = 31
                spe = 30
            }
            "Poison" -> {
                hp = 31
                atk = 31
                def = 30
                spa = 30
                spd = 30
                spe = 31
            }
            "Psychic" -> {
                hp = 30
                atk = 31
                def = 31
                spa = 31
                spd = 31
                spe = 30
            }
            "Rock" -> {
                hp = 31
                atk = 31
                def = 30
                spa = 31
                spd = 30
                spe = 30
            }
            "Steel" -> {
                hp = 31
                atk = 31
                def = 31
                spa = 31
                spd = 30
                spe = 31
            }
            "Water" -> {
                hp = 31
                atk = 31
                def = 31
                spa = 30
                spd = 31
                spe = 30
            }
        }
    }

    fun summaryText(nature: Nature = Nature.DEFAULT): CharSequence {
        val a = arrayOf(nature.plus, nature.minus)
        return (0 until 6).filter { i -> get(i) != 0 || a.contains(getName(i)!!.toId()) }
                .joinTo(SpannableStringBuilder()) { i ->
                    val name = getName(i)!!
                    val natSign = if (name.toId() == nature.plus) "+" else if (name.toId() == nature.minus) "-" else ""
                    "${get(i)}" concat  natSign.small()  concat name.small()
                }
    }

    companion object {

        @JvmStatic
        fun getName(index: Int) = when (index) {
            0 -> "HP"
            1 -> "Atk"
            2 -> "Def"
            3 -> "SpA"
            4 -> "Spd"
            5 -> "Spe"
            else -> null
        }

        @JvmStatic
        fun checkHpType(type: String?) = when (type?.trim()?.toLowerCase(Locale.ROOT)) {
            "bug" -> true
            "dark" -> true
            "dragon" -> true
            "electric" -> true
            "fighting" -> true
            "fire" -> true
            "flying" -> true
            "ghost" -> true
            "grass" -> true
            "ground" -> true
            "ice" -> true
            "poison" -> true
            "psychic" -> true
            "rock" -> true
            "steel" -> true
            "water" -> true
            else -> false
        }

        fun calculateSpeedRange(level: Int, baseSpe: Int, tier: String, gen: Int): IntArray {
            val isRandomBattle = tier.contains("Random Battle") || tier.contains("Random") && tier.contains("Battle") && gen >= 6
            val minNature = if (isRandomBattle || gen < 3) 1f else 0.9f
            val maxNature = if (isRandomBattle || gen < 3) 1f else 1.1f
            val maxIv = if (gen < 3) 30 else 31
            val min: Int
            var max: Int
            if (tier.contains("Let's Go")) {
                min = tr(tr(tr(2 * baseSpe * level / 100f + 5) * minNature) * tr((70f / 255f / 10f + 1) * 100) / 100f)
                max = tr(tr(tr((2 * baseSpe + maxIv) * level / 100f + 5) * maxNature) * tr((70f / 255f / 10f + 1) * 100) / 100f)
                if (tier.contains("No Restrictions")) max += 200 else if (tier.contains("Random")) max += 20
            } else {
                val maxIvEvOffset = maxIv + (if (isRandomBattle && gen >= 3) 21 else 63).toFloat()
                min = tr(tr(2 * baseSpe * level / 100f + 5) * minNature)
                max = tr(tr((2 * baseSpe + maxIvEvOffset) * level / 100f + 5) * maxNature)
            }
            return intArrayOf(min, max)
        }

        private fun tr(value: Float): Int {
            return value.toInt()
        }

        @JvmStatic
        fun calculateStat(base: Int, iv: Int, ev: Int, niv: Int, nat: Float): Int {
            return (((2 * base + iv + ev / 4) * niv / 100 + 5) * nat).toInt()
        }

        @JvmStatic
        fun calculateHp(base: Int, iv: Int, ev: Int, niv: Int): Int {
            return (2 * base + iv + ev / 4) * niv / 100 + niv + 10
        }

        fun toIndex(namee: String): Int {
            val name = namee.toLowerCase(Locale.ROOT).trim()
            if (name.contains("atk") || name.contains("attack")) {
                return if (name.contains("sp")) 3 else 1
            } else if (name.contains("def")) {
                return if (name.contains("sp")) 4 else 2
            } else if (name.contains("sp")) {
                return 5
            } else if (name.contains("h")) {
                return 0
            }
            return -1
        }
    }
}
